src/Azos/Data/ELink.cs

```diff
@@ -44,304 +42,295 @@ public sealed class ELink
+      if (m_Metadata != null && ((m_Metadata.Length * 2) > MAX_LINK_CHAR_SIZE))
+        throw new DataException(StringConsts.ELINK_CHAR_LENGTH_LIMIT_ERROR.Args("metadata[{0}]".Args(m_Metadata.Length)));
+    }
 
-                m_GDID = new GDID(era, id);
-                m_Metadata = metadata;
-            }
+    public ELink(string link)
+    {
+      if (link.IsNullOrWhiteSpace()) throw new DataException(StringConsts.ARGUMENT_ERROR + "ELink.ctor(link=null|empty)");
 
-        #endregion
+      if (link.Length > MAX_LINK_CHAR_SIZE)
+        throw new DataException(StringConsts.ELINK_CHAR_LENGTH_LIMIT_ERROR.Args(link.Substring(0, 20)));
 
+      m_Link = link;
+      decode();
     }
+
+    #endregion
+
+    #region Fields
+
+    private GDID m_GDID;
+    private byte[] m_Metadata;
+
+    private string m_Link;
+
+    #endregion
+
+    #region Properies
+
+    /// <summary>
+    /// Returns the ID portion of GDID represented by this instance
+    /// </summary>
+    public ulong ID { get { return m_GDID.ID; } }
+
+    /// <summary>
+    /// Returns metadata attached to this instance, or null if there is no metadata specified
+    /// </summary>
+    public byte[] Metadata { get { return m_Metadata; } }
+
+    /// <summary>
+    /// Returns a link encoded as a string using whatever randomization seed was passed to the last Encode(seed) call.
+    /// If Encode() was not called, then the link will get encoded using system rnd for a seed value
+    /// </summary>
+    public string Link
+    {
+      get
+      {
+        if (m_Link == null) Encode();
+        return m_Link;
+      }
+    }
+
+    /// <summary>
+    /// Returns the GDID that this link represents
+    /// </summary>
+    public GDID GDID { get { return m_GDID; } }
+
+    public GDIDSymbol AsGDIDSymbol { get { return new GDIDSymbol(m_GDID, Link); } }
+
+    #endregion
+
+    #region Methods
+
+    /// <summary>
+    /// Encodes a link into a textual form, using the supplied randomization seed, otherwise the system rnd is used.
+    /// A seed has 4 effective bits, yielding 16 possible variations for every link
+    /// </summary>
+    public string Encode(byte? seed = null) //props -> link
+    {
+
+      /* Format
+       * [lead1][lead2][csum] - [... era ...] - [... id ...]{-[...metadata...]}
+       *  1bt     1bt    1bt       0..4bts        1..8bts         md-len bts
+       *
+       * [lead1] = [rnd][authority]  rnd 0..f
+       *            4bi     4bi
+       * [lead2] = [eraLen][idLen]  idLen is w/o authority
+       *            4bi     4bi
+
+       * [csum] = CRC32(era, id, md_len, metadata_bytes) % 0xff
+       *
+       * [era] = LitEndian uint32 0..4 bytes
+       * [id] = LitEndian 1..8 bytes LitEndian uint64 w/o authority
+       *
+       *
+       * [metadata] = 0..limited by MAX_LINK_CHAR_SIZE/2
+       *
+       */
+
+      int rnd = (seed ?? (byte)(Platform.RandomGenerator.Instance.NextRandomInteger & 0xff)) & VARIATIONS;
```

Comment:
Use care with refactorings like this as the refactoring may not be equivalent in multi-threaded context. The instructions are NOT executed in the same order from a different CPU core perspective in the absence of memory barrier, therefore replacement of .HasValue with ?? is not ALWAYS equivalent for thread safety. It is better to keep these places as-is, although this specific case is not thread-safe anyway it is safer not to refactor these places and keep them as-is

Reply:
https://stackoverflow.com/questions/4619593/is-the-null-coalesce-operator-thread-safe